File: node/worker.js
/**
 * Consume and process tasks from an EZRA queue using a pool of async workers.
 *
 * Each worker maintains its own connection and unique identity. Parallel
 * processing is visible in the interleaved, color-coded log output.
 *
 * Scale containers with:
 *   docker compose up --scale worker=3
 *
 * Scale workers per container with:
 *   NUM_WORKERS=20 docker compose up
 *
 * Environment:
 *   EZRA_HOST      default: localhost
 *   EZRA_PORT      default: 42002
 *   EZRA_QUEUE     default: tasks
 *   NUM_WORKERS    parallel workers per container (default: 10)
 *   IDLE_TIMEOUT   seconds with no tasks before a worker exits (default: 3)
 *   SPEED          divide task duration by this; 5 = 5x faster (default: 5)
 *   NACK_MODE      immediate: nack with xdel, task retries right away (default)
 *                  deferred:  skip nack, EZRA reclaims after visibility_timeout seconds
 */

import os from 'os'
import { createClient } from 'redis'

const host        = process.env.EZRA_HOST    || 'localhost'
const port        = parseInt(process.env.EZRA_PORT    || '42002', 10)
const queue       = process.env.EZRA_QUEUE   || 'tasks'
const idleTimeout = parseInt(process.env.IDLE_TIMEOUT || '3',  10)
const speed       = parseFloat(process.env.SPEED      || '5')
const numWorkers  = parseInt(process.env.NUM_WORKERS  || '10', 10)
const nackMode    = process.env.NACK_MODE    || 'immediate'
const GROUP       = 'workers'
const BLOCK_MS    = 1000

const COLORS = [
  '\x1b[36m',  // cyan
  '\x1b[32m',  // green
  '\x1b[33m',  // yellow
  '\x1b[35m',  // magenta
  '\x1b[94m',  // bright blue
  '\x1b[96m',  // bright cyan
  '\x1b[92m',  // bright green
  '\x1b[93m',  // bright yellow
  '\x1b[95m',  // bright magenta
  '\x1b[34m',  // blue
]
const RESET = '\x1b[0m'
const BOLD  = '\x1b[1m'
const DIM   = '\x1b[2m'
const RED   = '\x1b[31m'

function log(color, label, action, detail = '') {
  console.log(`${color}[${label}]${RESET}  ${action.padEnd(5)}  ${detail}`)
}

const sleep = ms => new Promise(r => setTimeout(r, ms))

async function handleTask(task) {
  await sleep((parseFloat(task.duration || 1) / speed) * 1000)
  if (task.will_fail) throw new Error(`task ${task.id} is marked will_fail`)
}

async function runWorker(index) {
  const workerId = `${os.hostname()}-w${String(index).padStart(2, '0')}`
  const label    = `w-${String(index).padStart(2, '0')}`
  const color    = COLORS[(index - 1) % COLORS.length]

  const client = createClient({ socket: { host, port } })
  await client.connect()
  await client.clientSetName(workerId)

  try {
    await client.xGroupCreate(queue, GROUP, '$', { MKSTREAM: true })
  } catch (err) {
    if (!err.message.includes('BUSYGROUP')) throw err
  }

  log(color, label, 'ready')

  let done      = 0
  let nacked    = 0
  let idleSince = null

  while (true) {
    const response = await client.xReadGroup(
      GROUP, workerId,
      [{ key: queue, id: '>' }],
      { COUNT: 1, BLOCK: BLOCK_MS }
    )

    if (!response) {
      if (idleSince === null) idleSince = Date.now()
      if ((Date.now() - idleSince) / 1000 >= idleTimeout) {
        log(color, label, 'exit', `done=${done}  nacked=${nacked}`)
        break
      }
      continue
    }

    idleSince = null

    const { messages } = response[0]
    for (const { id: ezraId, message: fields } of messages) {
      const attempts    = fields.attempts     || '1'
      const maxAttempts = fields.max_attempts || '3'

      let task = {}
      try { task = JSON.parse(fields.payload || '{}') } catch { /* keep empty */ }

      const taskId   = String(task.id ?? '?').padStart(3)
      const duration = task.duration ?? '?'

      log(color, label, 'pop', `id=${taskId}  ${duration}s  ${attempts}/${maxAttempts}`)

      try {
        await handleTask(task)
        await client.xAck(queue, GROUP, ezraId)
        done++
        log(color, label, 'ack', `id=${taskId}`)
      } catch (err) {
        nacked++
        log(color, label, 'nack', `${RED}id=${taskId}  ${attempts}/${maxAttempts}  ${err.message}${RESET}`)
        if (nackMode === 'immediate') {
          await client.xDel(queue, ezraId)
        }
      }
    }
  }

  await client.disconnect()
  return { done, nacked }
}

// --- main ---

const hostname = os.hostname()
process.stdout.write(
  `\n${BOLD}ezra  ${RESET}${DIM}host=${hostname}  workers=${numWorkers}  queue=${queue}  ` +
  `idle_timeout=${idleTimeout}s  nack=${nackMode}${RESET}\n\n`
)

const results = await Promise.all(
  Array.from({ length: numWorkers }, (_, i) => runWorker(i + 1))
)

const totalDone   = results.reduce((s, r) => s + r.done,   0)
const totalNacked = results.reduce((s, r) => s + r.nacked, 0)

process.stdout.write(`\n${BOLD}done${RESET}  workers=${numWorkers}  tasks=${totalDone}  nacked=${totalNacked}\n`)
for (let i = 0; i < results.length; i++) {
  const r       = results[i]
  const color   = COLORS[i % COLORS.length]
  const label   = `w-${String(i + 1).padStart(2, '0')}`
  const nackStr = r.nacked ? `  ${RED}nacked=${r.nacked}${RESET}` : ''
  process.stdout.write(`  ${color}[${label}]${RESET}  done=${r.done}${nackStr}\n`)
}
process.stdout.write('\n')

File: python/worker.py
"""
Consume and process tasks from an EZRA queue using a pool of worker threads.

Each thread maintains its own connection and unique identity. Parallel
processing is visible in the interleaved, color-coded log output.

Scale containers with:
  docker compose up --scale worker=3

Scale threads per container with:
  NUM_WORKERS=20 docker compose up

Environment:
  EZRA_HOST      default: localhost
  EZRA_PORT      default: 42002
  EZRA_QUEUE     default: tasks
  NUM_WORKERS    parallel worker threads per container (default: 10)
  IDLE_TIMEOUT   seconds with no tasks before a worker exits (default: 30)
  NACK_MODE      immediate: nack with xdel, task retries right away (default)
                 deferred:  skip nack, EZRA reclaims after visibility_timeout seconds
"""

import json
import os
import socket
import time
import threading

import redis


QUEUE       = os.getenv("EZRA_QUEUE", "tasks")
GROUP       = "workers"
BLOCK_MS    = 1_000
NACK_MODE   = os.getenv("NACK_MODE", "immediate")
NUM_WORKERS = int(os.getenv("NUM_WORKERS", "10"))
SPEED       = float(os.getenv("SPEED", "1.0"))

_COLORS = [
    "\033[36m",  # cyan
    "\033[32m",  # green
    "\033[33m",  # yellow
    "\033[35m",  # magenta
    "\033[94m",  # bright blue
    "\033[96m",  # bright cyan
    "\033[92m",  # bright green
    "\033[93m",  # bright yellow
    "\033[95m",  # bright magenta
    "\033[34m",  # blue
]
_RESET = "\033[0m"
_BOLD  = "\033[1m"
_DIM   = "\033[2m"
_RED   = "\033[31m"

_print_lock = threading.Lock()


def _log(color: str, label: str, action: str, detail: str = "") -> None:
    with _print_lock:
        print(f"{color}[{label}]{_RESET}  {action:<5}  {detail}", flush=True)


def _connect(worker_id: str) -> redis.Redis:
    host = os.getenv("EZRA_HOST", "localhost")
    port = int(os.getenv("EZRA_PORT", "42002"))
    return redis.Redis(
        host=host, port=port,
        decode_responses=True, protocol=3,
        client_name=worker_id,
    )


def _ensure_group(client: redis.Redis) -> None:
    try:
        client.xgroup_create(QUEUE, GROUP, id="$", mkstream=True)
    except redis.ResponseError as exc:
        if "BUSYGROUP" not in str(exc):
            raise


def _process(task: dict) -> None:
    time.sleep(float(task.get("duration", 1.0)) / SPEED)
    if task.get("will_fail"):
        raise RuntimeError(f"task {task.get('id')} is marked will_fail")


def _run(index: int, idle_timeout: int, results: dict) -> None:
    hostname  = socket.gethostname()
    worker_id = f"{hostname}-w{index:02d}"
    label     = f"w-{index:02d}"
    color     = _COLORS[(index - 1) % len(_COLORS)]

    client = _connect(worker_id)
    _ensure_group(client)
    _log(color, label, "ready")

    done:   int         = 0
    nacked: int         = 0
    idle_since: float | None = None

    while True:
        response = client.xreadgroup(
            GROUP, worker_id, streams={QUEUE: ">"}, count=1, block=BLOCK_MS
        )

        if not response:
            if idle_since is None:
                idle_since = time.monotonic()
            elif time.monotonic() - idle_since >= idle_timeout:
                _log(color, label, "exit", f"done={done}  nacked={nacked}")
                break
            continue

        idle_since = None

        for _stream, (messages,) in response.items():
            for ezra_id, fields in messages:
                try:
                    task = json.loads(fields.get("payload", "{}"))
                except json.JSONDecodeError:
                    task = {}

                task_id      = task.get("id", "?")
                duration     = task.get("duration", "?")
                attempts     = fields.get("attempts", "1")
                max_attempts = fields.get("max_attempts", "3")

                _log(color, label, "pop",
                    f"id={str(task_id):>3}  {duration}s  {attempts}/{max_attempts}")

                try:
                    _process(task)
                    client.xack(QUEUE, GROUP, ezra_id)
                    done += 1
                    _log(color, label, "ack", f"id={str(task_id):>3}")

                except Exception as exc:
                    nacked += 1
                    _log(color, label, "nack",
                        f"{_RED}id={str(task_id):>3}  {attempts}/{max_attempts}  {exc}{_RESET}")
                    if NACK_MODE == "immediate":
                        client.xdel(QUEUE, ezra_id)

    results[index] = {"done": done, "nacked": nacked}
    client.close()


def main() -> None:
    idle_timeout = int(os.getenv("IDLE_TIMEOUT", "30"))
    hostname     = socket.gethostname()

    print(
        f"\n{_BOLD}ezra  {_RESET}{_DIM}host={hostname}  "
        f"workers={NUM_WORKERS}  queue={QUEUE}  "
        f"idle_timeout={idle_timeout}s  nack={NACK_MODE}{_RESET}\n",
        flush=True,
    )

    results: dict[int, dict] = {}
    threads = [
        threading.Thread(target=_run, args=(i, idle_timeout, results), daemon=True)
        for i in range(1, NUM_WORKERS + 1)
    ]

    for t in threads:
        t.start()

    for t in threads:
        t.join()

    total_done   = sum(r["done"]   for r in results.values())
    total_nacked = sum(r["nacked"] for r in results.values())

    print(f"\n{_BOLD}done{_RESET}  workers={NUM_WORKERS}  tasks={total_done}  nacked={total_nacked}")
    for i in sorted(results):
        r      = results[i]
        color  = _COLORS[(i - 1) % len(_COLORS)]
        label  = f"w-{i:02d}"
        nack_s = f"  {_RED}nacked={r['nacked']}{_RESET}" if r["nacked"] else ""
        print(f"  {color}[{label}]{_RESET}  done={r['done']}{nack_s}")
    print(flush=True)


if __name__ == "__main__":
    main()
